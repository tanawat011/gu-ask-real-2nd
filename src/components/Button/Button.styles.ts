import type { ButtonProps } from './Button'
import type { FlattenSimpleInterpolation } from 'styled-components'
import type { TwStyle } from 'twin.macro'
import type {
  ThemeMode,
  TwColor,
  TwColorLevel,
  TwShapeObject,
  TwSizeObject,
  Variant,
  WithRequired,
} from 'types'

import colors from 'tailwindcss/colors'
import tw, { styled, css } from 'twin.macro'

import { twColor } from 'utils/jest'

type TwThemeOption = {
  themeColor: TwColor
  colorLevel: TwColorLevel
  themeMode: ThemeMode
  isDisabled?: boolean
  isLoading?: boolean
}

type TwVariantFn = (
  variant: Variant,
  { themeColor, colorLevel, themeMode, isDisabled, isLoading }: TwThemeOption,
) => TwStyle | TwStyle[] | FlattenSimpleInterpolation

const twVariantFn: TwVariantFn = (
  variant,
  { themeColor, colorLevel, themeMode, isDisabled, isLoading },
) => {
  const isDarkMode = themeMode === 'dark'
  const color = colors[themeColor][colorLevel]
  const bgColor = twColor(color)
  const bgWhiteColor = twColor(colors.white)
  const borderColor = twColor(color, 'border')
  const textColor = twColor(color, 'text')
  const textWhiteColor = twColor(colors.white, 'text')
  const shadowColor = color

  const tertiaryBgColor = isDarkMode ? `${textWhiteColor}` : `${textColor}`

  if (isDisabled && !isLoading) {
    return tw`bg-gray-700 text-gray-100 opacity-50 hover:bg-gray-700`
  }

  switch (variant) {
    case 'secondary':
      return isDarkMode
        ? css`
            background-color: ${twColor(colors.gray[700])};
            border-width: 0;
            border-color: ${twColor(colors.gray[700], 'border')};
            color: ${twColor(colors.gray[100], 'text')};
            ${!isLoading &&
            `
              &:hover {
                background-color: ${twColor(colors.gray[600])};
                color: ${twColor(colors.gray[100], 'text')};
              }
              &:active {
                background-color: ${twColor(colors.gray[500])};
              }
            `}
          `
        : css`
            background-color: ${bgWhiteColor};
            border-width: 1px;
            border-color: ${twColor(colors.gray[300], 'border')};
            color: ${twColor(colors.gray[500], 'text')};
            ${!isLoading &&
            `
              &:hover {
                background-color: ${twColor(colors.gray[50])};
                color: ${twColor(colors.gray[500], 'text')};
              }
              &:active {
                background-color: ${twColor(colors.gray[100])};
              }
            `}
          `

    case 'tertiary':
      return css`
        --tw-bg-opacity: 0.2;
        background-color: ${bgColor};
        color: ${tertiaryBgColor};
        ${!isLoading &&
        `
          &:hover {
            --tw-bg-opacity: 0.3;
          }
          &:active {
            --tw-bg-opacity: 0.4;
          }
        `}
      `

    case 'outline':
      return css`
        --tw-border-opacity: 0.5;
        border-width: 2px;
        border-color: ${borderColor};
        color: ${textColor};
        ${!isLoading &&
        `
          &:hover {
            --tw-bg-opacity: 0.1;
            --tw-border-opacity: 1;
            background-color: ${bgColor};
          }
          &:active {
            --tw-border-opacity: 0.7;
          }
        `}
      `

    case 'shadow':
      return [
        tw`shadow-md`,
        css`
          --tw-bg-opacity: 1;
          --tw-shadow-color: ${shadowColor};
          --tw-shadow: var(--tw-shadow-colored);
          background-color: ${bgColor};
          color: ${textWhiteColor};
          ${!isLoading &&
          `
            &:hover {
              --tw-bg-opacity: 0.8;
              --tw-shadow-color: ${shadowColor}cc;
            }
            &:active {
              --tw-bg-opacity: 0.7;
              --tw-shadow-color: ${shadowColor}b3;
            }
          `}
        `,
      ]

    case 'text':
      return isDarkMode
        ? css`
            background-color: transparent;
            color: ${twColor(colors.gray[100], 'text')};
            ${!isLoading &&
            `
              &:hover {
                background-color: ${twColor(colors.gray[600])};
                color: ${twColor(colors.gray[100], 'text')};
              }
              &:active {
                background-color: ${twColor(colors.gray[500])};
              }
            `}
          `
        : css`
            background-color: transparent;
            color: ${twColor(colors.gray[500], 'text')};
            ${!isLoading &&
            `
              &:hover {
                background-color: ${twColor(colors.gray[50])};
                color: ${twColor(colors.gray[500], 'text')};
              }
              &:active {
                background-color: ${twColor(colors.gray[100])};
              }
            `}
          `

    case 'link':
      return tw`text-blue-500 hover:text-blue-600`

    case 'primary':
    default:
      return css`
        --tw-bg-opacity: 1;
        background-color: ${bgColor};
        color: ${textWhiteColor};
        ${!isLoading &&
        `
          &:hover {
            --tw-bg-opacity: 0.8;
          }
          &:active {
            --tw-bg-opacity: 0.7;
          }
        `}
      `
  }
}

const twSize: TwSizeObject = {
  xs: tw`h-7 px-3 py-1 text-xs`,
  sm: tw`h-9 px-3 py-2 text-sm`,
  md: tw`h-11 px-8 py-2 text-sm`,
  lg: tw`h-14 px-8 py-2 text-base`,
}

const twSizeIcon: TwSizeObject = {
  xs: tw`h-7 w-7 p-0 text-xs`,
  sm: tw`h-9 w-9 p-0 text-sm`,
  md: tw`h-11 w-11 p-0 text-sm`,
  lg: tw`h-14 w-14 p-0 text-base`,
}

const twShape: TwShapeObject = {
  square: tw`rounded-none`,
  rounded: tw`rounded-md`,
  circle: tw`rounded-full`,
}

type TwButtonProps = Omit<
  ButtonProps,
  'label' | 'icon' | 'onClick' | 'disabled'
> & {
  iconOnly?: boolean
  isDisabled?: boolean
  isLoading?: boolean
  isBlocked?: boolean
  themeColor: TwColor
  colorLevel: TwColorLevel
  themeMode: ThemeMode
}

export const TwSpan = tw.span`flex items-center justify-center gap-2`

export const TwButton = styled.button(
  ({
    variant,
    color,
    size,
    shape,
    iconOnly,
    isDisabled,
    isLoading,
    isBlocked,
    width,
    themeColor,
    colorLevel,
    themeMode,
  }: WithRequired<TwButtonProps, 'variant' | 'size' | 'shape'>) => {
    const [_color, _level] = (color?.split('-') as [TwColor, TwColorLevel]) || [
      themeColor,
      colorLevel,
    ]

    const variantOption = {
      themeColor: _color,
      colorLevel: _level,
      themeMode,
      isDisabled,
      isLoading,
    }

    return [
      twVariantFn(variant, variantOption),
      twShape[shape],
      iconOnly
        ? twSizeIcon[size]
        : [
            twSize[size],
            css`
              min-width: ${width};
            `,
          ],
      iconOnly && tw`rounded-full`,
      isDisabled && tw`cursor-not-allowed`,
      isLoading && tw`bg-opacity-50 text-opacity-50`,
      isBlocked && tw`w-full`,
      tw`relative flex select-none items-center justify-center`,
    ]
  },
)
