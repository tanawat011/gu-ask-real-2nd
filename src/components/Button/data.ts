import type { TwShapeObject, TwSizeObject, TwVariantObject } from 'types'

import tw from 'twin.macro'

import {
  TwBgDanger,
  TwBgDark,
  TwBgInfo,
  TwBgLight,
  TwBgPrimary,
  TwBgSecondary,
  TwBgSuccess,
  TwBgTertiary,
  TwBgWarning,
  TwBorderDanger,
  TwBorderDark,
  TwBorderInfo,
  TwBorderLight,
  TwBorderPrimary,
  TwBorderSecondary,
  TwBorderSuccess,
  TwBorderTertiary,
  TwBorderWarning,
  TwTextDanger,
  TwTextDark,
  TwTextInfo,
  TwTextLight,
  TwTextPrimary,
  TwTextSecondary,
  TwTextSuccess,
  TwTextTertiary,
  TwTextWarning,
} from 'components/Variant'

export const twVariant: { bg: TwVariantObject; border: TwVariantObject } = {
  bg: {
    primary: [TwBgPrimary, tw`text-white`],
    secondary: [TwBgSecondary, tw`text-white`],
    tertiary: [TwBgTertiary, tw`text-white`],
    success: [TwBgSuccess, tw`text-white`],
    danger: [TwBgDanger, tw`text-white`],
    warning: [TwBgWarning, tw`text-white`],
    info: [TwBgInfo, tw`text-white`],
    light: [TwBgLight, tw`text-gray-800`],
    dark: [TwBgDark, tw`text-white`],
    link: tw`text-blue-500 hover:text-blue-600`,
  },
  border: {
    primary: [TwBorderPrimary, TwTextPrimary, tw`border-2`],
    secondary: [TwBorderSecondary, TwTextSecondary, tw`border-2`],
    tertiary: [TwBorderTertiary, TwTextTertiary, tw`border-2`],
    success: [TwBorderSuccess, TwTextSuccess, tw`border-2`],
    danger: [TwBorderDanger, TwTextDanger, tw`border-2`],
    warning: [TwBorderWarning, TwTextWarning, tw`border-2`],
    info: [TwBorderInfo, TwTextInfo, tw`border-2`],
    light: [TwBorderLight, TwTextLight, tw`border-2`],
    dark: [TwBorderDark, TwTextDark, tw`border-2`],
    link: tw`text-blue-500 hover:text-blue-600`,
  },
}

export const twSize: TwSizeObject = {
  xs: tw`h-7 px-3 py-1 text-xs`,
  sm: tw`h-9 px-3 py-2 text-sm`,
  md: tw`h-11 px-8 py-2 text-sm`,
  lg: tw`h-14 px-8 py-2 text-base`,
  xl: tw`h-20 px-8 py-2 text-lg`,
}

export const twShape: TwShapeObject = {
  none: tw`rounded-none`,
  rounded: tw`rounded-md`,
  circle: tw`rounded-full`,
}
