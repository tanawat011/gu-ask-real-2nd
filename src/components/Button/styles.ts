import type { TwShapeObject, TwSizeObject, TwVariantObject } from 'types'

import tw from 'twin.macro'

import {
  TwBgDanger,
  TwBgDark,
  TwBgInfo,
  TwBgLight,
  TwBgPlain,
  TwBgPrimary,
  TwBgSecondary,
  TwBgSuccess,
  TwBgTertiary,
  TwBgWarning,
  TwBorderDanger,
  TwBorderDark,
  TwBorderInfo,
  TwBorderLight,
  TwBorderPlain,
  TwBorderPrimary,
  TwBorderSecondary,
  TwBorderSuccess,
  TwBorderTertiary,
  TwBorderWarning,
  TwTextDanger,
  TwTextDark,
  TwTextInfo,
  TwTextLight,
  TwTextPlain,
  TwTextPrimary,
  TwTextSecondary,
  TwTextSuccess,
  TwTextTertiary,
  TwTextWarning,
} from 'components/Variant'

export const twVariant: { bg: TwVariantObject; border: TwVariantObject } = {
  bg: {
    primary: [TwBgPrimary, tw`text-white`],
    secondary: [TwBgSecondary, tw`text-white`],
    tertiary: [TwBgTertiary, tw`text-white`],
    success: [TwBgSuccess, tw`text-white`],
    danger: [TwBgDanger, tw`text-white`],
    warning: [TwBgWarning, tw`text-white`],
    info: [TwBgInfo, tw`text-white`],
    light: [TwBgLight, tw`text-gray-800`],
    dark: [TwBgDark, tw`text-white`],
    plain: [TwBgPlain, tw`text-white`],
    link: tw`text-blue-500 hover:text-blue-600`,
  },
  border: {
    primary: [TwBorderPrimary, TwTextPrimary, tw`border-2`],
    secondary: [TwBorderSecondary, TwTextSecondary, tw`border-2`],
    tertiary: [TwBorderTertiary, TwTextTertiary, tw`border-2`],
    success: [TwBorderSuccess, TwTextSuccess, tw`border-2`],
    danger: [TwBorderDanger, TwTextDanger, tw`border-2`],
    warning: [TwBorderWarning, TwTextWarning, tw`border-2`],
    info: [TwBorderInfo, TwTextInfo, tw`border-2`],
    light: [TwBorderLight, TwTextLight, tw`border-2`],
    dark: [TwBorderDark, TwTextDark, tw`border-2`],
    plain: [TwBorderPlain, TwTextPlain, tw`border-2`],
    link: tw`text-blue-500 hover:text-blue-600`,
  },
}

export const twSize: TwSizeObject = {
  xs: tw`h-7 px-3 py-1 text-xs`,
  sm: tw`h-9 px-3 py-2 text-sm`,
  md: tw`h-11 px-8 py-2 text-sm`,
  lg: tw`h-14 px-8 py-2 text-base`,
  xl: tw`h-20 px-8 py-2 text-lg`,
}

export const twSizeIcon: TwSizeObject = {
  xs: tw`h-7 w-7 p-0 text-xs`,
  sm: tw`h-9 w-9 p-0 text-sm`,
  md: tw`h-11 w-11 p-0 text-sm`,
  lg: tw`h-14 w-14 p-0 text-base`,
  xl: tw`h-20 w-20 p-0 text-lg`,
}

export const twShape: TwShapeObject = {
  none: tw`rounded-none`,
  rounded: tw`rounded-md`,
  circle: tw`rounded-full`,
}
