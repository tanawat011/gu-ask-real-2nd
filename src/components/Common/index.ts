export * from './FlexCol'
export * from './FlexRow'
export * from './Card'
