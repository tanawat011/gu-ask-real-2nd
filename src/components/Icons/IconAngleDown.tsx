import type { SizeProp } from '@fortawesome/fontawesome-svg-core'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

type IconAngleDownProps = {
  size?: SizeProp
}

export const IconAngleDown: React.FC<IconAngleDownProps> = ({ size }) => {
  return <FontAwesomeIcon icon='angle-down' size={size} />
}
