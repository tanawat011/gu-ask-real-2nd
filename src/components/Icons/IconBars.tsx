import type { SizeProp } from '@fortawesome/fontawesome-svg-core'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

type IconBarsProps = {
  size?: SizeProp
}

export const IconBars: React.FC<IconBarsProps> = ({ size }) => {
  return <FontAwesomeIcon icon='bars' size={size} />
}
