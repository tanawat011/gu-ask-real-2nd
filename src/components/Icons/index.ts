export * from './IconAngleDown'
export * from './IconBars'
export * from './IconBell'
export * from './IconCaretLeft'
export * from './IconCode'
export * from './IconCopy'
export * from './IconGear'
export * from './IconHome'
export * from './IconHouse'
export * from './IconLoadingBar'
export * from './IconLoadingBarCenter'
export * from './IconLoadingBarCenterSequence'
export * from './IconLoadingBarFullSequence'
export * from './IconLoadingBarTop'
export * from './IconLoadingCircle'
export * from './IconLoadingCircle2'
export * from './IconLoadingSign'
export * from './IconSpinner'
export * from './IconUser'
