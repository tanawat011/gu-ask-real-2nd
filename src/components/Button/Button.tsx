import type { Variant, Size, Shape } from 'types'

import type { MouseEvent } from 'react'

import tw, { styled } from 'twin.macro'

import { Loading } from 'components/Loading'

import { twShape, twSize, twSizeIcon, twVariant } from './styles'

type ButtonProps = {
  label?: string
  icon?: React.ReactNode
  iconR?: React.ReactNode
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void
  variant?: Variant
  size?: Size
  outline?: boolean
  shape?: Shape
  disabled?: boolean
  loading?: boolean
  block?: boolean
}

type TwButtonProps = Omit<ButtonProps, 'label' | 'icon' | 'onClick' | 'disabled'> & {
  iconOnly?: boolean
  disabledOnly?: boolean
  isDisabled?: boolean
  isLoading?: boolean
  isBlocked?: boolean
}

const TwContainer = tw.div`relative`
const TwButton = styled.button(
  ({
    variant,
    size,
    outline,
    shape,
    iconOnly,
    disabledOnly,
    isDisabled,
    isLoading,
    isBlocked,
  }: TwButtonProps) => {
    return [
      variant && (outline ? twVariant.border[variant] : twVariant.bg[variant]),
      size && (iconOnly ? twSizeIcon[size] : twSize[size]),
      shape && twShape[shape],
      iconOnly && tw`rounded-full`,
      isDisabled && tw`cursor-not-allowed select-none`,
      disabledOnly && tw`(bg-charcoal text-anti-flash-white opacity-40)!`,
      isLoading && tw`text-opacity-30 opacity-50!`,
      isBlocked && tw`w-full`,
      tw`relative`,
    ]
  },
)
const TwSpan = tw.span`flex items-center justify-center gap-2`

export const Button: React.FC<ButtonProps> = ({
  label,
  icon,
  iconR,
  onClick,
  variant = 'primary',
  size = 'md',
  outline,
  shape = 'rounded',
  disabled,
  loading,
  block,
}) => {
  const handleOnClick = (event: MouseEvent<HTMLButtonElement>) => {
    event.preventDefault()

    onClick && onClick(event)
  }

  const isIconOnly = !!icon && !label
  const disabledOnly = disabled && !loading
  const isDisabled = disabled || loading

  return (
    <TwContainer>
      <TwButton
        onClick={handleOnClick}
        variant={variant}
        size={size}
        outline={outline}
        iconOnly={isIconOnly}
        shape={shape}
        disabled={isDisabled}
        disabledOnly={disabledOnly}
        isDisabled={isDisabled}
        isLoading={loading}
        isBlocked={block}
      >
        {isIconOnly ? (
          icon
        ) : (
          <>
            <TwSpan>
              {icon}
              {label}
              {iconR}
            </TwSpan>
          </>
        )}
      </TwButton>

      {loading && <Loading />}
    </TwContainer>
  )
}
