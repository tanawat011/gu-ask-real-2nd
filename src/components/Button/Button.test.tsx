/* eslint-disable jest/no-conditional-expect */
import type { Size, Variant } from 'types'

import { fireEvent, screen } from '@testing-library/react'

import { IconHome } from 'components/Icons'
import { renderWithProviders, twColor } from 'utils/jest'

import { Button } from './Button'

jest.mock('./Button', () => jest.requireActual('./Button'))

describe('<Button />', () => {
  test('renders correctly', async () => {
    expect.assertions(4)

    const onClick = jest.fn()

    const {
      asFragment,
      container: { firstChild },
    } = renderWithProviders(<Button label='button' onClick={onClick} />)

    expect(firstChild).toHaveStyle('position: relative')
    expect(firstChild).toHaveStyle('user-select: none')
    expect(onClick).toHaveBeenCalledTimes(0)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders correctly and handle event `onClick`', async () => {
    expect.assertions(1)

    const onClick = jest.fn()

    renderWithProviders(<Button label='button' onClick={onClick} />)

    const button = screen.getByRole('button')
    fireEvent.click(button)

    expect(onClick).toHaveBeenCalledTimes(1)
  })

  test('renders correctly and disabled', async () => {
    expect.assertions(6)

    renderWithProviders(<Button label='button' disabled />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyleRule('background-color', `${twColor('#293341')} !important`)
    expect(button).toHaveStyleRule('color', `${twColor('#6a717a', 'text')} !important`)
    expect(button).toHaveStyle('cursor: not-allowed')
    expect(button).toHaveAttribute('disabled')
  })

  test('renders correctly and loading', async () => {
    expect.assertions(5)

    renderWithProviders(<Button label='button' loading />)

    const button = screen.getByRole('button')
    const loadingIcon = screen.getByText('Loading')

    expect(button).toHaveStyle('--tw-text-opacity: 0.3')
    expect(button).toHaveStyle('opacity: 0.5')
    expect(button).toHaveStyle('cursor: not-allowed')
    expect(button).toHaveAttribute('disabled')
    expect(loadingIcon).toBeInTheDocument()
  })

  test('renders correctly and blocked', async () => {
    expect.assertions(1)

    renderWithProviders(<Button label='button' block />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyle('width: 100%')
  })

  test('renders correctly and display icon only', async () => {
    expect.assertions(1)

    renderWithProviders(<Button icon={<IconHome />} />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyle('border-radius: 9999px')
  })

  type SizeOption = {
    height: string
    paddingX: string
    paddingY: string
    fontSize: string
    lineHeight: string
  }

  type TestSize = { [key in Size]: SizeOption }

  const sizes: TestSize = {
    xs: {
      fontSize: '0.75rem',
      lineHeight: '1rem',
      height: '1.75rem',
      paddingX: '0.75rem',
      paddingY: '0.25rem',
    },
    sm: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      height: '2.25rem',
      paddingX: '0.75rem',
      paddingY: '0.5rem',
    },
    md: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      height: '2.75rem',
      paddingX: '2rem',
      paddingY: '0.5rem',
    },
    lg: {
      fontSize: '1rem',
      lineHeight: '1.5rem',
      height: '3.5rem',
      paddingX: '2rem',
      paddingY: '0.5rem',
    },
  }

  Object.entries(sizes).forEach(([key, { fontSize, lineHeight, height, paddingX, paddingY }]) => {
    test(`renders correctly and size \`${key}\``, () => {
      expect.assertions(7)

      renderWithProviders(<Button label='button' size={key as Size} />)

      const button = screen.getByRole('button')

      expect(button).toHaveStyle(`height: ${height}`)
      expect(button).toHaveStyle(`padding-top: ${paddingY}`)
      expect(button).toHaveStyle(`padding-bottom: ${paddingY}`)
      expect(button).toHaveStyle(`padding-left: ${paddingX}`)
      expect(button).toHaveStyle(`padding-right: ${paddingX}`)
      expect(button).toHaveStyle(`font-size: ${fontSize}`)
      expect(button).toHaveStyle(`line-height: ${lineHeight}`)
    })
  })

  type SizeIconOption = {
    widthHeight: string
    fontSize: string
    lineHeight: string
  }

  type TestSizeIcon = { [key in Size]: SizeIconOption }

  const sizesIcon: TestSizeIcon = {
    xs: {
      fontSize: '0.75rem',
      lineHeight: '1rem',
      widthHeight: '1.75rem',
    },
    sm: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      widthHeight: '2.25rem',
    },
    md: {
      fontSize: '0.875rem',
      lineHeight: '1.25rem',
      widthHeight: '2.75rem',
    },
    lg: {
      fontSize: '1rem',
      lineHeight: '1.5rem',
      widthHeight: '3.5rem',
    },
  }

  Object.entries(sizesIcon).forEach(([key, { fontSize, lineHeight, widthHeight }]) => {
    test(`renders correctly and display icon only with size \`${key}\``, () => {
      expect.assertions(6)

      renderWithProviders(<Button icon={<IconHome />} size={key as Size} />)

      const button = screen.getByRole('button')

      expect(button).toHaveStyle('border-radius: 9999px')
      expect(button).toHaveStyle(`height: ${widthHeight}`)
      expect(button).toHaveStyle(`width: ${widthHeight}`)
      expect(button).toHaveStyle('padding: 0')
      expect(button).toHaveStyle(`font-size: ${fontSize}`)
      expect(button).toHaveStyle(`line-height: ${lineHeight}`)
    })
  })

  type VariantOption = {
    bgColor?: string
    bgHoverColor?: string
    textColor?: string
    textHoverColor?: string
  }

  type TestVariant = { [key in Variant]: VariantOption }

  const variants: TestVariant = {
    primary: {
      bgColor: '#4F46E5',
      textColor: '#ffffff',
    },
    secondary: {
      bgColor: '#374151',
      textColor: '#ffffff',
    },
    tertiary: {
      bgColor: '#2d355c',
      textColor: '#ffffff',
    },
    solid: {
      bgColor: '#15803d',
      textColor: '#ffffff',
    },
    twoTone: {
      bgColor: '#b91c1c',
      textColor: '#ffffff',
    },
    plain: {
      bgHoverColor: '#4B5563',
      textColor: '#ffffff',
    },
    link: {
      textHoverColor: '#2563EB',
      textColor: '#3B82F6',
    },
  }

  Object.entries(variants).forEach(
    ([key, { bgColor, bgHoverColor, textColor, textHoverColor }]) => {
      test(`renders correctly and variant \`${key}\``, async () => {
        expect.hasAssertions()

        renderWithProviders(<Button label='button' variant={key as Variant} />)

        const button = screen.getByRole('button')

        if (bgColor) {
          expect(button).toHaveStyleRule('background-color', twColor(bgColor))
        }

        if (bgHoverColor) {
          expect(button).toHaveStyleRule('background-color', twColor(bgHoverColor), {
            modifier: ':hover',
          })
        }

        if (textColor) {
          expect(button).toHaveStyleRule('color', twColor(textColor, 'text'))
        }

        if (textHoverColor) {
          expect(button).toHaveStyleRule('color', twColor(textHoverColor, 'text'), {
            modifier: ':hover',
          })
        }

        if (!(['plain', 'link'] as Variant[]).includes(key as Variant)) {
          expect(button).toHaveStyleRule('--tw-bg-opacity', '0.8', { modifier: ':hover' })
        }
      })
    },
  )

  type OutlineVariantOption = {
    borderColor?: string
    borderHoverColor?: string
    textColor?: string
    textHoverColor?: string
  }

  type TestOutlineVariant = { [key in Variant]: OutlineVariantOption }

  const outlineVariants: TestOutlineVariant = {
    primary: {
      borderColor: '#4F46E5',
    },
    secondary: {
      borderColor: '#374151',
    },
    tertiary: {
      borderColor: '#2d355c',
    },
    solid: {
      borderColor: '#15803d',
    },
    twoTone: {
      borderColor: '#b91c1c',
    },
    plain: {
      borderColor: 'transparent',
      borderHoverColor: '#4B5563',
    },
    link: {
      textColor: '#3B82F6',
      textHoverColor: '#2563EB',
    },
  }

  Object.entries(outlineVariants).forEach(
    ([key, { borderColor, borderHoverColor, textColor, textHoverColor }]) => {
      test(`renders correctly and outline variant \`${key}\``, () => {
        expect.hasAssertions()

        renderWithProviders(<Button label='button' variant={key as Variant} outline />)

        const button = screen.getByRole('button')

        if (borderColor) {
          expect(button).toHaveStyleRule(
            'border-color',
            borderColor === 'transparent' ? borderColor : twColor(borderColor, 'border'),
          )
          expect(button).toHaveStyleRule(
            'color',
            borderColor === 'transparent' ? borderColor : twColor(borderColor, 'text'),
          )
        }

        if (borderHoverColor) {
          expect(button).toHaveStyleRule('border-color', twColor(borderHoverColor, 'border'), {
            modifier: ':hover',
          })
          expect(button).toHaveStyleRule('color', twColor(borderHoverColor, 'text'), {
            modifier: ':hover',
          })
        }

        if (key !== 'link') {
          expect(button).toHaveStyle('border-width: 2px')
        }

        if (textColor && textHoverColor) {
          expect(button).toHaveStyleRule('color', twColor(textColor, 'text'))
          expect(button).toHaveStyleRule('color', twColor(textHoverColor, 'text'), {
            modifier: ':hover',
          })
        }

        if (!(['plain', 'link'] as Variant[]).includes(key as Variant)) {
          expect(button).toHaveStyleRule('--tw-border-opacity', '0.8', { modifier: ':hover' })
          expect(button).toHaveStyleRule('--tw-text-opacity', '0.8', { modifier: ':hover' })
        }
      })
    },
  )

  test('renders correctly and shape square', () => {
    expect.hasAssertions()

    renderWithProviders(<Button label='button' shape='square' />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyle('border-radius: 0px')
  })

  test('renders correctly and shape rounded', () => {
    expect.hasAssertions()

    renderWithProviders(<Button label='button' shape='rounded' />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyle('border-radius: 0.375rem')
  })

  test('renders correctly and shape circle', () => {
    expect.hasAssertions()

    renderWithProviders(<Button label='button' shape='circle' />)

    const button = screen.getByRole('button')

    expect(button).toHaveStyle('border-radius: 9999px')
  })
})
