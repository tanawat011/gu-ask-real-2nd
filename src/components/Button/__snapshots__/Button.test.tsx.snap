// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  --tw-bg-opacity: 1;
  background-color: rgb(79 70 229 / var(--tw-bg-opacity));
  color: rgb(255 255 255 / var(--tw-text-opacity));
  height: 2.75rem;
  padding-left: 2rem;
  padding-right: 2rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  border-radius: 0.375rem;
}

.c1:hover {
  --tw-bg-opacity: 0.8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 0.5rem;
}

<div
    class="c0"
  >
    <button
      class="c1"
      shape="rounded"
    >
      <span
        class="c2"
      >
        button
      </span>
    </button>
  </div>
</DocumentFragment>
`;
